fendermint/vm/interpreter/src/fvm/gas/actor.rs

```diff
@@ -1,123 +1,105 @@
-/// is used. However, the calling pattern should be single threaded, so direct `store` could be used.
-/// The usage of `Atomic` is purely to bypass the compilation issue without using unsafe.
-/// TODO: remove this overhead when trait bound is updated.
 #[derive(Default)]
 pub struct ActorGasMarket {
     /// The block gas limit
-    block_gas_limit: AtomicGas,
+    block_gas_limit: Gas,
     /// The accumulated gas usage so far
-    block_gas_used: AtomicGas,
+    block_gas_used: Gas,
 }
 
 impl GasMarket for ActorGasMarket {
     fn available_block_gas(&self) -> Gas {
-        self.block_gas_limit.load(Ordering::SeqCst) - self.block_gas_used.load(Ordering::SeqCst)
+        self.block_gas_limit - self.block_gas_used
     }
 
-    fn consume_gas(&self, gas: Gas) -> anyhow::Result<()> {
-        let block_gas_used = self.block_gas_used.load(Ordering::SeqCst);
-
-        if block_gas_used + gas >= self.block_gas_limit.load(Ordering::SeqCst) {
-            anyhow::bail!("out of block gas")
+    fn record_gas_used(&mut self, gas: Gas) -> anyhow::Result<()> {
+        if self.block_gas_used + gas >= self.block_gas_limit {
+            tracing::warn!("out of block gas, should not have happened")
         }
-        self.block_gas_used
-            .store(block_gas_used + gas, Ordering::SeqCst);
+        self.block_gas_used = self.block_gas_used.saturating_add(gas);
 
         Ok(())
     }
+}
 
-    fn update_params<DB: Blockstore + Clone + 'static>(
-        &self,
-        chain_state: &mut FvmExecState<DB>,
-    ) -> anyhow::Result<()> {
-        let block_gas_used = self.block_gas_used.load(Ordering::SeqCst);
-        let params = fvm_ipld_encoding::RawBytes::serialize(block_gas_used)?;
-
+impl ActorGasMarket {
+    pub fn new<E: Executor>(
+        executor: &mut E,
+        block_height: ChainEpoch,
+    ) -> anyhow::Result<ActorGasMarket> {
         let msg = FvmMessage {
             from: system::SYSTEM_ACTOR_ADDR,
-            to: GAS_ACTOR_ADDR,
-            sequence: chain_state.block_height() as u64,
+            to: GAS_MARKET_ACTOR_ADDR,
+            sequence: block_height as u64,
             // exclude this from gas restriction
-            gas_limit: u64::MAX,
-            method_num: fendermint_actor_gas_market::Method::UpdateBlockGasConsumption as u64,
-            params,
+            gas_limit: fvm_shared::BLOCK_GAS_LIMIT,
+            method_num: fendermint_actor_gas_market::Method::CurrentGasReading as u64,
+            params: fvm_ipld_encoding::RawBytes::default(),
             value: Default::default(),
             version: Default::default(),
             gas_fee_cap: Default::default(),
             gas_premium: Default::default(),
         };
 
-        let (apply_ret, _) = chain_state.execute_implicit(msg)?;
+        let raw_length = fvm_ipld_encoding::to_vec(&msg).map(|bz| bz.len())?;
+        let apply_ret = executor.execute_message(msg, ApplyKind::Implicit, raw_length)?;
```

Comment:
I wish we could perform this call with `readonly`, as it would result in better safety guarantees. It's not possible with the current FVM APIs, right?

Reply:
yeah, I dont see any readonly execution, but I think we can pass the store as readonly for some txns